### PR TITLE
TimePicker

### PR DESCRIPTION
**Pull request by atomiomi**

### CODE REVIEW

File: src/time/index.es [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import React, { useEffect, useState, forwardRef } from 'react';
+
+import Time from './time';
+import Input from '../input';
+
+
+const mobileDeviceRegex = /iPad|iPhone|iPod|Android|Windows Phone/;
```
Comment (suprMax) [inline src/time/index.es]: -> switch to our hook

File: src/time/time_input.es [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      ref={inputRef}
+      type="text"
+      maxLength={5}
+      onChange={handleChange}
+      onKeyDown={handleKeyDown}
+    />
+  );
+};
+
+const WrappedTimeInput = forwardRef(TimeInput);
```
Comment (suprMax) [inline src/time/time_input.es]: wrap on export as usual?

File: src/time/time_input.es [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    const cleanedValue = cleanValue(value);
+
+    if (!value) return invoke(onChange, value);
+    if (!isInt(cleanedValue)) return;
+
+    const { selectionEnd } = inputRef.current;
+    const formattedValue = formatValue(cleanedValue);
+
+    // Change value in DOM to restore caret position
+    inputRef.current.value = formattedValue;
```
Comment (suprMax) [inline src/time/time_input.es]: why not use state with a controlled input?

File: src/time/utils.es [hunk 4]
```diff
@@ -0,0 +1,31 @@
+export const cleanValue = (value) => value.replace(/:/g, '');
```
Comment (suprMax) [inline src/time/utils.es]: why not simple ':' - we only gonna have 1

File: src/time/utils.es [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+export const getCaretPosition = (currentPosition, oldValue, newValue) => {
+  const isCaretAtEnd = currentPosition >= oldValue.length;
+  return isCaretAtEnd ? newValue.length : currentPosition;
+};
+
+const limitHours = (value) => Math.min(23, Math.max(value, 0));
+const limitMinutes = (value) => Math.min(59, Math.max(value, 0));
+
+const transformValue = (value, caretPosition, transform) => {
```
Comment (suprMax) [inline src/time/utils.es]: take caret position out of this and split fns?

**Comment by suprMax** [comment]
1. Allow writing `:`
2. padLeft on arrow increment/decrement
3. min/max demo in form
4. placeholder demo

File: src/time/time_input.es (hunk 2 above)
Comment (atomiomi) [reply src/time/time_input.es:68]: can't define defaultProps on func that is passed to forward ref

File: src/time/utils.es (hunk 4 above)
Comment (atomiomi) [reply src/time/utils.es:1]: user can paste any input

**Comment by suprMax** [comment]
@atomiomi  Check validations, the demo is still broken

**Comment by suprMax** [comment]
@atomiomi don't merge yet, we need to demo this and get a "go-ahead"

**Pull request merged**